typing/subst.ml

```diff
@@ -432,149 +501,267 @@ let rename_bound_idents scoping s sg =
-    mtd_attributes = attrs s decl.mtd_attributes;
-    mtd_loc = loc s decl.mtd_loc;
-    mtd_uid = decl.mtd_uid;
-  }
-
-
-(* For every binding k |-> d of m1, add k |-> f d to m2
-   and return resulting merged map. *)
-
-let merge_path_maps f m1 m2 =
-  Path.Map.fold (fun k d accu -> Path.Map.add k (f d) accu) m1 m2
-
-let keep_latest_loc l1 l2 =
-  match l2 with
-  | None -> l1
-  | Some _ -> l2
-
-let type_replacement s = function
-  | Path p -> Path (type_path s p)
-  | Type_function { params; body } ->
-    For_copy.with_scope (fun copy_scope ->
-     let params = List.map (typexp copy_scope s) params in
-     let body = typexp copy_scope s body in
-     Type_function { params; body })
+and subst_lazy_signature_item' copy_scope scoping s comp =
+  match comp with
+    SigL_value(id, d, vis) ->
+      SigL_value(id, value_description' copy_scope s d, vis)
+  | SigL_type(id, d, rs, vis) ->
+      SigL_type(id, type_declaration' copy_scope s d, rs, vis)
+  | SigL_typext(id, ext, es, vis) ->
+      SigL_typext(id, extension_constructor' copy_scope s ext, es, vis)
+  | SigL_module(id, pres, d, rs, vis) ->
+      SigL_module(id, pres, subst_lazy_module_decl scoping s d, rs, vis)
+  | SigL_modtype(id, d, vis) ->
+      SigL_modtype(id, subst_lazy_modtype_decl scoping s d, vis)
+  | SigL_class(id, d, rs, vis) ->
+      SigL_class(id, class_declaration' copy_scope s d, rs, vis)
+  | SigL_class_type(id, d, rs, vis) ->
+      SigL_class_type(id, cltype_declaration' copy_scope s d, rs, vis)
+
+and force_signature_item = function
+  | SigL_value(id, vd, vis) -> Sig_value(id, vd, vis)
+  | SigL_type(id, d, rs, vis) -> Sig_type(id, d, rs, vis)
+  | SigL_typext(id, ext, es, vis) -> Sig_typext(id, ext, es, vis)
+  | SigL_module(id, pres, d, rs, vis) ->
+     Sig_module(id, pres, force_module_decl d, rs, vis)
+  | SigL_modtype(id, d, vis) ->
+     Sig_modtype (id, force_modtype_decl d, vis)
+  | SigL_class(id, d, rs, vis) -> Sig_class(id, d, rs, vis)
+  | SigL_class_type(id, d, rs, vis) -> Sig_class_type(id, d, rs, vis)
+
+and modtype scoping s t =
+  t |> lazy_modtype |> subst_lazy_modtype scoping s |> force_modtype
 
 (* Composition of substitutions:
      apply (compose s1 s2) x = apply s2 (apply s1 x) *)
 
-let compose s1 s2 =
+and compose s1 s2 =
+  if s1 == identity then s2 else
+  if s2 == identity then s1 else
   { types = merge_path_maps (type_replacement s2) s1.types s2.types;
     modules = merge_path_maps (module_path s2) s1.modules s2.modules;
     modtypes = merge_path_maps (modtype Keep s2) s1.modtypes s2.modtypes;
     for_saving = s1.for_saving || s2.for_saving;
     loc = keep_latest_loc s1.loc s2.loc;
   }
+
+
+let subst_lazy_signature_item scoping s comp =
+  For_copy.with_scope
+    (fun copy_scope -> subst_lazy_signature_item' copy_scope scoping s comp)
+
+let module_declaration scoping s decl =
+  decl
+  |> lazy_module_decl |> subst_lazy_module_decl scoping s |> force_module_decl
```

Comment:
This should probably be moved down with the other similar functions below `Lazy` and adjusted to match their style.